File: backend/app/utils/formatters.py
from datetime import datetime, timedelta
from typing import Optional, Any, Dict, List
import json
import re

def format_time_ago(date: datetime) -> str:
    if not date:
        return "unknown"
    
    now = datetime.utcnow()
    delta = now - date
    
    seconds = delta.total_seconds()
    minutes = seconds / 60
    hours = minutes / 60
    days = hours / 24
    weeks = days / 7
    months = days / 30
    years = days / 365
    
    if seconds < 60:
        return "just now"
    elif minutes < 60:
        return f"{int(minutes)}m ago"
    elif hours < 24:
        return f"{int(hours)}h ago"
    elif days < 7:
        return f"{int(days)}d ago"
    elif weeks < 4:
        return f"{int(weeks)}w ago"
    elif months < 12:
        return f"{int(months)}mo ago"
    else:
        return f"{int(years)}y ago"

def format_date_short(date: datetime) -> str:
    if not date:
        return "N/A"
    return date.strftime("%b %d, %Y")

def format_date_full(date: datetime) -> str:
    if not date:
        return "N/A"
    return date.strftime("%b %d, %Y %H:%M")

def format_time_iso(date: datetime) -> str:
    if not date:
        return None
    return date.isoformat() + "Z"

def format_number(num: int) -> str:
    if num is None:
        return "0"
    
    if num < 1000:
        return str(num)
    elif num < 1000000:
        return f"{num/1000:.1f}K"
    elif num < 1000000000:
        return f"{num/1000000:.1f}M"
    else:
        return f"{num/1000000000:.1f}B"

def format_percentage(value: float, decimals: int = 1) -> str:
    if value is None:
        return "0%"
    return f"{value * 100:.{decimals}f}%"

def format_currency(amount: float, currency: str = "$") -> str:
    if amount is None:
        return f"{currency}0.00"
    return f"{currency}{amount:,.2f}"

def format_file_size(size_bytes: int) -> str:
    if size_bytes < 1024:
        return f"{size_bytes} B"
    elif size_bytes < 1024**2:
        return f"{size_bytes/1024:.1f} KB"
    elif size_bytes < 1024**3:
        return f"{size_bytes/1024**2:.1f} MB"
    elif size_bytes < 1024**4:
        return f"{size_bytes/1024**3:.1f} GB"
    else:
        return f"{size_bytes/1024**4:.1f} TB"

def format_confidence(score: float) -> str:
    if score is None:
        return "0%"
    return f"{score * 100:.1f}%"

def get_confidence_level(score: float) -> str:
    if score >= 0.85:
        return "high"
    elif score >= 0.70:
        return "medium"
    elif score >= 0.55:
        return "low"
    else:
        return "negative"

def get_confidence_color(score: float) -> str:
    level = get_confidence_level(score)
    colors = {
        "high": "#4ade80",     
        "medium": "#facc15",   
        "low": "#fb923c",     
        "negative": "#f87171", 
    }
    return colors.get(level, "#94a3b8") 

def get_platform_icon(platform: str) -> str:
    icons = {
        "instagram": "📸",
        "facebook": "👍",
        "twitter": "🐦",
        "tiktok": "🎵",
        "telegram": "✈️",
        "reddit": "🤖",
        "youtube": "▶️",
        "linkedin": "💼",
        "snapchat": "👻",
        "pinterest": "📌",
    }
    return icons.get(platform.lower(), "🌐")

def get_platform_color(platform: str) -> str:
    colors = {
        "instagram": "#E4405F",
        "facebook": "#1877F2",
        "twitter": "#1DA1F2",
        "tiktok": "#000000",
        "telegram": "#26A5E4",
        "reddit": "#FF4500",
        "youtube": "#FF0000",
        "linkedin": "#0A66C2",
        "snapchat": "#FFFC00",
        "pinterest": "#E60023",
    }
    return colors.get(platform.lower(), "#94a3b8")

def pretty_json(data: Any) -> str:
    try:
        return json.dumps(data, indent=2, default=str)
    except:
        return str(data)

def safe_json_parse(json_str: str) -> Optional[Dict]:
    try:
        return json.loads(json_str)
    except:
        return None

def truncate_text(text: str, max_length: int = 100, suffix: str = "...") -> str:
    if not text:
        return ""
    
    if len(text) <= max_length:
        return text
    
    return text[:max_length - len(suffix)] + suffix

def truncate_middle(text: str, max_length: int = 100, suffix: str = "...") -> str:
    if not text:
        return ""
    
    if len(text) <= max_length:
        return text
    
    half = (max_length - len(suffix)) // 2
    return text[:half] + suffix + text[-half:]

__all__ = [
    "format_time_ago",
    "format_date_short",
    "format_date_full",
    "format_time_iso",
    
    "format_number",
    "format_percentage",
    "format_currency",
    
    "format_file_size",
    
    "format_confidence",
    "get_confidence_level",
    "get_confidence_color",
    
    "get_platform_icon",
    "get_platform_color",
    
    "pretty_json",
    "safe_json_parse",
    
    "truncate_text",
    "truncate_middle",
]

File: client/src/utils/formatters.ts
import { formatDistanceToNow, format, parseISO, isValid } from "date-fns";

export const formatTimeAgo = (date: string | Date): string => {
    if (!date) return 'unknown';
    try {
        const parsed = typeof date === 'string' ? parseISO(date) : date;
        if (isValid(parsed)) {
            return formatDistanceToNow(parsed, { addSuffix: true });
        }
    } catch {
        return 'invalid date';
    }
    return 'unknown'
};

export const formatDateShort = (date: string | Date): string => {
    if (!date) return 'N/A';
    try {
        const parsed = typeof date === 'string' ? parseISO(date) : date;
        if (isValid(parsed)) {
            return format(parsed, 'MMM dd, yyyy');
        }
    } catch {
        return 'invalid date'
    }
    return 'N/A'
};

export const formatDateFull = (date: string | Date): string => {
    if (!date) return 'N/A'
    try {
        const parsed = typeof date === 'string' ? parseISO(date) : date;
        if (isValid(parsed)) {
            return format(parsed, 'MMMM dd, yyyy HH:mm:ss');
        }
    } catch {
        return 'invalid date'
    }
    return 'N/A'
};

export const formatDateISO = (date: string | Date): string => {
    if (!date) return 'N/A'
    try {
        const parsed = typeof date === 'string' ? parseISO(date) : date;
        if (isValid(parsed)) {
            return parsed.toISOString();
        }
    } catch {
        return '';
    }
    return '';
};

export const formatDuration = (ms: number): string => {
    if (ms < 1000) return `${ms.toFixed(0)}ms`;
    const seconds = Math.floor(ms / 1000);
    if (seconds < 60) return `${seconds}s`;
    const minutes = Math.floor(seconds / 60);
    const remainingSeconds = seconds % 60;
    if (minutes < 60) return `${minutes}m ${remainingSeconds}s`;
    const hours = Math.floor(minutes / 60);
    const remainingMinutes = minutes % 60;
    if (hours < 24) return `${hours}h ${remainingMinutes}m`;
    const days = Math.floor(hours / 24);
    const remainingHours = hours % 24;
    return `${days}d ${remainingHours}h`;
};

export const formatNumber = (num: number | null | undefined): string => {
    if (num == null) return '0';
    if (num < 1000) return num.toString();
    if (num < 1_000_000) return `${(num / 1_000).toFixed(0)}K`;
    if (num < 1_000_000_000) return `${(num / 1_000_000).toFixed(1)}M`;
    if (num < 1_000_000_000_000) return `${(num / 1_000_000_000).toFixed(1)}B`;
    return `${(num / 1_000_000_000_000).toFixed(1)}T`;
};

export const formatCurrency = (amount: number, currency = 'KSH'): string => {
    if (amount == null) return `${currency}0.00`;
    return `${currency}${amount.toLocaleString('en-US', {
        minimumFractionDigits: 2,
        maximumFractionDigits: 2,
    })}`;
};

export const formatPercentage = (value: number, decimals = 1): string => {
    if (value == null) return '0%';
    return `${(value * 100).toFixed(decimals)}%`;
};

export const formatBytes = (bytes: number, decimals = 2): string => {
    if (bytes === 0) return '0 B';
    const k = 1024;
    const sizes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(decimals))} ${sizes[i]}`;
};

export const formatFileSize = formatBytes;


export const formatConfidence = (score: number | null | undefined): string => {
    if (score == null) return '0%';
    return `${(score * 100).toFixed(1)}%`;
};

export const getConfidenceLevel = (score: number): 'high' | 'medium' | 'low' | 'negative' => {
    if (score >= 0.85) return 'high';
    if (score >= 0.70) return 'medium';
    if (score >= 0.55) return 'low';
    return 'negative';
};

export const getConfidenceColor = (score: number): string => {
    const level = getConfidenceLevel(score);
    const colors = {
        high: '#4ade80',
        medium: '#facc15',
        low: '#fb923c',
        negative: '#f87171',
    };
    return colors[level];
};


const PLATFORM_ICONS: Record<string, string> = {
    instagram: '📸',
    facebook: '👍',
    twitter: '🐦',
    tiktok: '🎵',
    telegram: '✈️',
    reddit: '🤖',
    youtube: '▶️',
    linkedin: '💼',
    snapchat: '👻',
    pinterest: '📌',
    discord: '🎮',
    whatsapp: '💬',
};

const PLATFORM_COLORS: Record<string, string> = {
    instagram: '#E4405F',
    facebook: '#1877F2',
    twitter: '#1DA1F2',
    tiktok: '#000000',
    telegram: '#26A5E4',
    reddit: '#FF4500',
    youtube: '#FF0000',
    linkedin: '#0A66C2',
    snapchat: '#FFFC00',
    pinterest: '#E60023',
    discord: '#5865F2',
    whatsapp: '#25D366',
};

export const getPlatformIcon = (platform: string): string =>
    PLATFORM_ICONS[platform.toLowerCase()] || '🌐';

export const getPlatformColor = (platform: string): string =>
    PLATFORM_COLORS[platform.toLowerCase()] || '#94a3b8';


export const prettyJson = (data: any): string => {
    try {
        return JSON.stringify(data, null, 2);
    } catch {
        return String(data);
    }
};

export const safeJsonParse = <T = any>(jsonStr: string): T | null => {
    try {
        return JSON.parse(jsonStr);
    } catch {
        return null;
    }
};

export const truncateText = (text: string, maxLength = 100, suffix = '...'): string => {
    if (!text) return '';
    if (text.length <= maxLength) return text;
    return text.slice(0, maxLength - suffix.length) + suffix;
};

export const truncateMiddle = (text: string, maxLength = 100, suffix = '...'): string => {
    if (!text) return '';
    if (text.length <= maxLength) return text;
    const half = Math.floor((maxLength - suffix.length) / 2);
    return text.slice(0, half) + suffix + text.slice(-half);
};

export const formatPhone = (phone: string): string => {
    const cleaned = phone.replace(/\D/g, '');
    if (cleaned.length === 10) {
        return `(${cleaned.slice(0, 3)}) ${cleaned.slice(3, 6)}-${cleaned.slice(6)}`;
    }
    return phone;
};

export const formatUrl = (url: string, maxLength = 50): string => {
    if (!url) return '';
    try {
        const parsed = new URL(url);
        const display = `${parsed.hostname}${parsed.pathname}`;
        return truncateMiddle(display, maxLength);
    } catch {
        return truncateMiddle(url, maxLength);
    }
};